joomla-framework/string
# Portable UTF-8

Pull Request for Issue #29

This pull request replaces the abandoned library `phputf8` with **Portable UTF-8**.

> `Portable UTF-8` is written in PHP and _can_ work without `mbstring`, `iconv` or any other extra encoding php-extension on the server. It is easy to use and provides much more functionality than the former implementation. This library will 
auto-detect the server environment and will use the installed php-extensions if they are available, 
so the best possible performance is provided. As a fallback, [Symfony Polyfills](https://github.com/symfony/polyfill) are used, if needed.

(Freely quoted from [Portable UTF-8](https://github.com/voku/portable-utf8/blob/master/README.md))

This PR is kept separate from the Test Consolidation #37 for now, so the tests can be updated and used for both new and old implementation, if issues arise.

### Summary of Changes

* Replaced `phputf8` with `Portable UTF-8`
* Revised inline documentation
* Added tests to verify that signature consolidation does not break existing code

### Testing Instructions

* Code review. Check especially that
  * signatures did not change in a breaking way
  * behaviour does not change in an unexpected way

### Documentation Changes Required

Yes, but should be generated from source. Inline documentation has been revised.

### Outstanding Decisions

1. Do we consider the utf8_* functions to be part of this package's API, as Micheal suggests?
2. Do we want to encourage the use of an external library (thus deprecate proxy methods calling that library) or do we want to keep the whole scope of services and maintain it ourselves?
